core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go

```diff
@@ -39,52 +41,75 @@ type EVMAutomationUpkeepResult21 struct {
+}
+
+func (enc EVMAutomationEncoder21) Encode(results ...ocr2keepers.CheckResult) ([]byte, error) {
+	if len(results) == 0 {
+		return nil, ErrEmptyResults
 	}
 
 	report := automation_utils_2_1.KeeperRegistryBase21Report{
-		UpkeepIds:    make([]*big.Int, len(toReport)),
-		GasLimits:    make([]*big.Int, len(toReport)),
-		Triggers:     make([][]byte, len(toReport)),
-		PerformDatas: make([][]byte, len(toReport)),
+		UpkeepIds:    make([]*big.Int, len(results)),
+		GasLimits:    make([]*big.Int, len(results)),
+		Triggers:     make([][]byte, len(results)),
+		PerformDatas: make([][]byte, len(results)),
 	}
 
-	for i, result := range toReport {
-		res, ok := result.(EVMAutomationUpkeepResult21)
+	for i, result := range results {
+		ext, ok := result.Extension.(EVMAutomationResultExtension21)
 		if !ok {
-			return nil, fmt.Errorf("unexpected upkeep result struct")
+			return nil, fmt.Errorf("unexpected check result extension struct")
 		}
 
 		// only take these values from the first result
 		// TODO: find a new way to get these values
 		if i == 0 {
-			report.FastGasWei = res.FastGasWei
-			report.LinkNative = res.LinkNative
+			report.FastGasWei = ext.FastGasWei
+			report.LinkNative = ext.LinkNative
+		}
+
+		id, ok := big.NewInt(0).SetString(string(result.Payload.Upkeep.ID), 10)
+		if !ok {
+			return nil, fmt.Errorf("failed to parse big int from upkeep id: %s", string(result.Payload.Upkeep.ID))
 		}
 
-		report.UpkeepIds[i] = res.ID
-		report.GasLimits[i] = res.GasUsed
-		trigger, err := enc.packer.PackTrigger(res.ID, triggerWrapper{
-			BlockNum:  res.CheckBlockNumber,
-			BlockHash: res.CheckBlockHash,
-			// TODO: fill with real info
-			// LogIndex: 0,
-			// TxHash:   [32]byte{},
-		})
+		report.UpkeepIds[i] = id
+		report.GasLimits[i] = big.NewInt(0).SetUint64(result.GasAllocated)
+
+		triggerW := triggerWrapper{
+			BlockNum:  uint32(result.Payload.Trigger.BlockNumber),
+			BlockHash: common.HexToHash(result.Payload.Trigger.BlockHash),
+		}
+		switch getUpkeepType(id.Bytes()) {
+		case logTrigger:
+			trExt, ok := result.Payload.Trigger.Extension.(logprovider.LogTriggerExtension)
+			if !ok {
+				return nil, fmt.Errorf("unrecognized trigger extension data")
+			}
+			hex, err := common.ParseHexOrString(trExt.TxHash)
+			if err != nil {
+				return nil, fmt.Errorf("tx hash parse error: %w", err)
+			}
+			triggerW.TxHash = common.BytesToHash(hex[:])
+			triggerW.LogIndex = uint32(trExt.LogIndex)
+		default:
+		}
+		trigger, err := enc.packer.PackTrigger(id, triggerW)
 		if err != nil {
 			return nil, fmt.Errorf("%w: failed to pack trigger", err)
 		}
 		report.Triggers[i] = trigger
-		report.PerformDatas[i] = res.PerformData
+		report.PerformDatas[i] = result.PerformData
 	}
 
 	return enc.packer.PackReport(report)
 }
 
-func (enc EVMAutomationEncoder21) DecodeReport(raw []byte) ([]ocr2keepers.UpkeepResult, error) {
+func (enc EVMAutomationEncoder21) Decode(raw []byte) ([]ocr2keepers.UpkeepResult, error) {
```

Comment:
I'm not sure we use `Decode` anywhere anymore as the intent was to replace it with `Extract` since we don't use `UpkeepResult` in the new version. You should be able to delete this.

Reply:
Currently it is used from packer.UnpackTransmitTxInput within `TransmitEventProvider` when fetching check block number for events that are parsed.

Reply:
this was done in 2.0 since events did not have checkBlockNumber logged. It should no longer be needed in 2.1 as they should have the triggerID as a logged field